File: hbackup.py
#!/usr/bin/python3

import socket
import sys
import os
import datetime
import time
import urllib.parse
import hashlib
import argparse
import re

debug = False
haserror = False
total_files = total_dirs = total_links = skipped_files = total_file_len = upload_file_len = 0
exclude_file_name_patterns = []
md5sum_cache_file = None
md5sum_cache = {}

#md5sum_cache = { filename: [st_mtime, md5sum], ... }
#md5sum_cache_file
#len md5sum file_name


def load_md5sum_cache(md5sum_cache_file):
    global md5sum_cache
    BUFSIZE = 1024
    print("reading md5sum_cache_file ...")
    f = open(md5sum_cache_file, 'r')
    cnt = 0
    try:
        lines = f.readlines(BUFSIZE)
        while lines:
            for line in lines:
                c = line.split(' ')
                if len(c) != 3:
                    print("error in md5sum_cache_file: " + line + "\n")
                    sys.exit(-1)
                c[2] = c[2].strip("\n")
                v = [int(c[0]), c[1]]
                md5sum_cache.update({c[2]: v})
                cnt += 1
            lines = f.readlines(BUFSIZE)
    finally:
        f.close()
    print("loaded md5sum_cache %d lines" % cnt)


def save_md5sum_cache(md5sum_cache_file):
    print("saveing md5sum_cache_file ...")
    f = open(md5sum_cache_file, "w")
    c = 0
    for key, value in md5sum_cache.items():
        f.write("%d %s %s\n" % (value[0], value[1], key))
        c += 1
    f.close()
    print("saved md5sum_cache %d lines" % c)


def md5sum(filename, blocksize=1024 * 1024):
    global md5sum_cache
    if md5sum_cache_file != None:
        v = md5sum_cache.get(filename)
        if v != None:
            if int(os.stat(filename).st_mtime) == v[0]:
                if debug:
                    print("using md5sumcache")
                return v[1]
            else:
                if debug:
                    print("file st_mtime %d, md5sum_cache st_mtime %d" % (int(
                        os.stat(filename).st_mtime), v[0]))
        else:
            if debug:
                print("not in md5sum_cache")

    hash = hashlib.md5()
    with open(filename, "rb") as f:
        for block in iter(lambda: f.read(blocksize), b""):
            hash.update(block)
    md5sum_str = hash.hexdigest()
    if md5sum_cache_file != None:
        v = [int(os.stat(filename).st_mtime), md5sum_str]
        md5sum_cache.update({filename: v})
    if debug:
        print("update md5sum_cache")
    return md5sum_str


def end_hbackup(retcode=0):
    s.send('END\n'.encode())
    data = s.recv(100).decode()
    print('S', data, end='')
    print('FDL: %d/%d/%d, skipped %d, U/A: %d/%d' %
          (total_files, total_dirs, total_links, skipped_files,
           upload_file_len, total_file_len))
    if md5sum_cache_file != None:
        save_md5sum_cache(md5sum_cache_file)
    if haserror:
        print("Encountered error when backuping file")
        print("Error msg append to " + err_log + ", please check it")
    sys.exit(retcode)


def send_dir(remote_name):
    global total_dirs
    total_dirs += 1
    s.send(('MKDIR ' + urllib.parse.quote(remote_name) + '\n').encode())
    data = s.recv(100).decode()
    if data[0:2] == 'OK':
        print("")
        if debug:
            print('S', data, end='')
        return
    print(' S', data, end='')
    log_err(remote_name + ' ' + data)


def send_link(remote_name, linkto):
    global total_links
    total_links += 1
    s.send(('MKLINK ' + urllib.parse.quote(remote_name) + ' ' +
            urllib.parse.quote(linkto) + '\n').encode())
    data = s.recv(100).decode()
    if data[0:2] == 'OK':
        print("")
        if debug:
            print('S', data, end='')
        return
    print(' S', data, end='')
    log_err(remote_name + ' ' + data)


def send_file(local_file_name, remote_name):
    global total_files, total_file_len, upload_file_len
    total_files += 1
    filemd5sum = md5sum(local_file_name)
    file_size = os.path.getsize(local_file_name)
    total_file_len += file_size
    if debug:
        print(filemd5sum + "_" + str(file_size))
    s.send(('FILE ' + filemd5sum + ' ' + str(file_size) + ' ' +
            urllib.parse.quote(remote_name) + '\n').encode())
    data = s.recv(100).decode()
    if data[0:2] == 'OK':
        if debug:
            print('S', data, end='')
        print("")
        return
    if data[0:5] == 'ERROR':
        print(' S', data, end='')
        log_err(local_file_name + ' --> ' + remote_name + ' ' + data)
        return
    if data[0:4] == 'DATA':
        if debug:
            print('S', data, end='')
            print('I will send file')
        CHUNKSIZE = 1024 * 1024
        file = open(local_file_name, "rb")
        bytes_send = 0
        try:
            while True:
                need_read = file_size - bytes_send
                if need_read > 0:
                    if need_read > CHUNKSIZE:
                        bytes_read = file.read(CHUNKSIZE)
                    else:
                        bytes_read = file.read(need_read)
                    if len(bytes_read) > 0:
                        upload_file_len += len(bytes_read)
                        s.send(bytes_read)
                        bytes_send += len(bytes_read)
                    else:
                        break
                else:
                    break
        finally:
            file.close()
        data = s.recv(100).decode()
        if data[0:2] == 'OK':
            if debug:
                print('S', data, end='')
            print("")
            return
    print('S', data, end='')
    end_hbackup(-1)


def log_err(msg):
    global haserror
    haserror = True
    if err_log == "":
        print(msg)
        exit(-1)
    f = open(err_log, 'a')
    now = datetime.datetime.now()
    f.write(str(now) + ' ' + msg)
    f.close()


parser = argparse.ArgumentParser(description='hbackup')
parser.add_argument(dest='hostname', metavar='HostName')
parser.add_argument(dest='port', metavar='TcpPort', type=int)
parser.add_argument(dest='password', metavar='Password')
parser.add_argument(dest='file_name', metavar='File/DirToSend')
parser.add_argument(dest='remote_name', metavar='RemoteName', nargs='?')
parser.add_argument(
    '-x',
    dest='exclude_file_name',
    action='append',
    metavar='exclude_file_regex',
    help='exclude_file_name_regex')
parser.add_argument('-d', dest='debug', action='store_true', help='debug mode')
parser.add_argument(
    '-t',
    dest='days',
    metavar='n',
    action='store',
    help='skip n days old files',
    type=int)
parser.add_argument(
    '-e',
    dest='err_log',
    metavar='err_log_file',
    action='store',
    help='error msg will be append to err_log_file, and continue to run')
parser.add_argument(
    '-m',
    dest='md5sum_cache_file',
    metavar='md5sum_cache_file',
    action='store',
    help='md5sum_cache_file')
args = parser.parse_args()
err_log = args.err_log
if err_log == None:
    err_log = ""
debug = args.debug
host = args.hostname
port = args.port
pass_word = args.password
file_name = args.file_name
if args.exclude_file_name == None:
    exclude_file_name_patterns = []
else:
    for exname in args.exclude_file_name:
        print("exclude: " + exname)
        exclude_file_name_patterns.append(re.compile(exname))

if args.remote_name == None:
    file_new_name = file_name
else:
    file_new_name = args.remote_name

if args.days == None:
    file_mtime_start = None
else:
    file_mtime_start = time.mktime(
        (datetime.datetime.now() -
         datetime.timedelta(days=args.days)).timetuple())
    print("skip file mtime before " + datetime.datetime.fromtimestamp(
        file_mtime_start).strftime('%Y-%m-%d %H:%M:%S'))

if args.md5sum_cache_file == None:
    md5sum_cache_file = None
else:
    md5sum_cache_file = args.md5sum_cache_file
    load_md5sum_cache(args.md5sum_cache_file)

try:
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    s.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
except socket.error:
    print('Failed to creat socket.')
    sys.exit(-1)
try:
    host_ip = socket.gethostbyname(host)
except socket.gaierror:
    print('Host name could not be resolved. Exiting...')
    sys.exit(-1)

try:
    s.connect((host_ip, port))
except socket.error:
    if s:
        s.close()
    print('Socket connection is not established!\t' + message)
    sys.exit(1)
if debug:
    print('Connected to ' + host + ' on IP ' + host_ip + ' port ' + str(port) +
          '.')

s.send(('PASS ' + pass_word + '\n').encode())
data = s.recv(100).decode()
if data[0:2] != 'OK':
    if debug:
        print('S', data, end='')
    sys.exit(-1)

if os.path.islink(file_name):
    print(file_name + " is symlink", end='')
    linkto = os.readlink(file_name)
    send_link(file_new_name, linkto)
    end_hbackup()

if os.path.isfile(file_name):
    print(file_name + " is file", end='')
    send_file(file_name, file_new_name)
    end_hbackup()

if not os.path.isdir(file_name):
    print(file_name + " is not symlink, file, dir, I do not know how to deal")
    end_hbackup()

print(file_name + " is dir")

for root, dirs, files in os.walk(file_name, topdown=True):
    for name in files:
        local_file_name = os.path.join(root, name)
        if exclude_file_name_patterns:
            skip = False
            for expatterns in exclude_file_name_patterns:
                if expatterns.match(local_file_name) != None:
                    print(local_file_name + " SKIP excluded")
                    skip = True
                    break
            if skip:
                skipped_files += 1
                continue
        if file_mtime_start != None:
            if os.lstat(local_file_name).st_mtime < file_mtime_start:
                print(local_file_name + " SKIP old file")
                skipped_files += 1
                continue
        remote_file_name = file_new_name + '/' + root[len(file_name) +
                                                      1:] + '/' + name
        if os.sep == "\\":
            remote_file_name = remote_file_name.replace("\\", "/")
        if debug:
            print("F root=" + root + " name=" + name + " file_new_name=" +
                  file_new_name)
            print(local_file_name + " --> " + remote_file_name)
        if os.path.islink(local_file_name):
            print(local_file_name + " is symlink", end='')
            linkto = os.readlink(local_file_name)
            send_link(remote_file_name, linkto)
        elif os.path.isfile(local_file_name):
            print(local_file_name, end='')
            if debug:
                print(" is file")
            send_file(local_file_name, remote_file_name)
        else:
            print(local_file_name, " SKIP")

    for name in dirs:
        local_file_name = os.path.join(root, name)
        if exclude_file_name_patterns:
            skip = False
            for expatterns in exclude_file_name_patterns:
                if expatterns.match(local_file_name) != None:
                    print(local_file_name + " SKIP excluded")
                    skip = True
                    break
            if skip:
                skipped_files += 1
                continue
        if file_mtime_start != None:
            if os.lstat(local_file_name).st_mtime < file_mtime_start:
                print(local_file_name + " SKIP old file")
                skipped_files += 1
                continue
        remote_file_name = file_new_name + '/' + root[len(file_name) +
                                                      1:] + '/' + name
        if os.path.islink(local_file_name):
            print(local_file_name + " is symlink", end='')
            linkto = os.readlink(local_file_name)
            send_link(remote_file_name, linkto)
            continue
        print(os.path.join(root, name) + " is dir", end='')
        if os.sep == "\\":
            remote_file_name = remote_file_name.replace("\\", "/")
        send_dir(remote_file_name)

end_hbackup()

File: hbackup.c
/* hbackup
	  by james@ustc.edu.cn 2018.12.29
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <fcntl.h>
#include <unistd.h>
#include <syslog.h>
#include <signal.h>
#include <sys/ioctl.h>
#include <sys/time.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <time.h>
#include <pwd.h>
#include <net/if.h>
#include <netinet/ip.h>
#include <netinet/ip6.h>
#include <netinet/tcp.h>
#include <netdb.h>
#include <stdarg.h>
#include <errno.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <pthread.h>
#include <openssl/md5.h>
#include <linux/limits.h>
#include <libgen.h>
#include <dirent.h>

#include "uthash.h"

#include "util.c"

char md5cache_file[PATH_MAX];
FILE *md5cache_fp;

char error_log_file[PATH_MAX];
FILE *error_log_fp;

char local_file_name[PATH_MAX];
char remote_file_name[PATH_MAX];

int haserror = 0;
size_t total_files, total_dirs, total_links, skipped_files, total_file_len, upload_file_len;

struct my_struct {
	const char *filename;	/* key */
	time_t filetime;
	int used;
	char md5s[MD5_DIGEST_LENGTH * 2 + 1];
	UT_hash_handle hh;	/* makes this structure hashable */
};

struct my_struct *md5sum_cache = NULL;
int md5sum_cache_lookup, md5sum_cache_hit, md5sum_cache_update;

void update_md5sum_cache(char *name, time_t ft, char *md5s)
{
	struct my_struct *s = NULL;
	char *p;
	if (md5cache_fp == NULL)
		return;
	HASH_FIND_STR(md5sum_cache, name, s);
	if (s) {
		if (debug)
			printf("update md5sum_cache %lu %s %s\n", ft, md5s, name);
		s->filetime = ft;
		s->used = 1;
		strncpy(s->md5s, md5s, MD5_DIGEST_LENGTH * 2);
		return;
	}
	s = (struct my_struct *)malloc(sizeof(struct my_struct));
	p = malloc(strlen(name) + 1);
	if (p == NULL) {
		printf("malloc error\n");
		exit(-1);
	}
	strcpy(p, name);
	s->filename = p;
	s->filetime = ft;
	s->used = 1;
	strncpy(s->md5s, md5s, MD5_DIGEST_LENGTH * 2);
	HASH_ADD_KEYPTR(hh, md5sum_cache, s->filename, strlen(s->filename), s);
	if (debug)
		printf("add md5sum_cache %lu %s %s\n", ft, md5s, name);
}

void load_md5sum_cache()
{
	char buf[MAXLEN];
	int cnt = 0;

	if (md5cache_fp == NULL)
		return;
	printf("loading md5sum_cache from %s ...", md5cache_file);
	while (fgets(buf, MAXLEN, md5cache_fp)) {
		char *p1, *p2, *p3;
		p1 = buf;
		p2 = buf;
		while (*p2 && (*p2 != ' '))
			p2++;
		if (*p2 == 0) {
			printf("skip %s", buf);
			continue;
		}
		p3 = p2 + 1;
		while (*p3 && (*p3 != ' '))
			p3++;
		if (*p3 == 0) {
			printf("skip %s", buf);
			continue;
		}
		*p2 = 0;
		p2++;
		*p3 = 0;
		p3++;
		if (*p3 == 0)
			continue;
		if (p3[strlen(p3) - 1] == '\n')
			p3[strlen(p3) - 1] = 0;
		time_t ft;
		if (sscanf(p1, "%lu", &ft) != 1)
			continue;
		if (strlen(p2) != MD5_DIGEST_LENGTH * 2)
			continue;
		if (strlen(p3) <= 0)
			continue;
		if (debug)
			printf("cache %lu %s %s\n", ft, p2, p3);
		update_md5sum_cache(p3, ft, p2);
		cnt += 1;
	}
	printf("loaded md5sum_cache %d lines\n", cnt);
}

void save_md5sum_cache()
{
	if (md5cache_fp == NULL)
		return;
	rewind(md5cache_fp);
	ftruncate(fileno(md5cache_fp), 0);
	struct my_struct *s;

	for (s = md5sum_cache; s != NULL; s = s->hh.next) {
		if (s->used)
			fprintf(md5cache_fp, "%lu %s %s\n", s->filetime, s->md5s, s->filename);
	}
	fclose(md5cache_fp);
}

time_t file_mtime(char *fname)
{
	struct stat buf;
	if (stat(fname, &buf) == 0)
		return buf.st_mtime;
	return 0;
}

char *file_md5sum(char *fname)
{
	time_t ft;
	if (md5cache_fp) {
		struct my_struct *s;
		ft = file_mtime(fname);
		HASH_FIND_STR(md5sum_cache, fname, s);
		if (s) {
			md5sum_cache_lookup++;
			if (s->filetime == ft) {
				s->used = 1;
				md5sum_cache_hit++;
				return s->md5s;
			}
		}
	}
	char *md5s;
	md5s = get_file_md5sum(fname);
	if (md5cache_fp) {
		md5sum_cache_update++;
		update_md5sum_cache(fname, ft, md5s);
	}
	return md5s;
}

void log_err(const char *fmt, ...)
{
	va_list ap;
	va_start(ap, fmt);
	haserror = 1;
	if (error_log_fp) {
		fprintf(error_log_fp, "%s ", stamp());
		vfprintf(error_log_fp, fmt, ap);
		va_end(ap);
		fflush(error_log_fp);
	} else {
		vprintf(fmt, ap);
		va_end(ap);
		exit(-1);
	}
}

int tcp_connect(const char *host, const char *serv)
{
	int sockfd, n;
	struct addrinfo hints, *res, *ressave;

	bzero(&hints, sizeof(struct addrinfo));
	hints.ai_family = AF_UNSPEC;
	hints.ai_socktype = SOCK_STREAM;

	if ((n = getaddrinfo(host, serv, &hints, &res)) != 0)
		err_quit("tcp_connect error for %s, %s", host, serv);
	ressave = res;

	do {
		sockfd = socket(res->ai_family, res->ai_socktype, res->ai_protocol);
		if (sockfd < 0)
			continue;	/* ignore this one */

		if (connect(sockfd, res->ai_addr, res->ai_addrlen) == 0)
			break;	/* success */

		close(sockfd);	/* ignore this one */
	} while ((res = res->ai_next) != NULL);

	if (res == NULL)	/* errno set from final connect() */
		err_sys("tcp_connect error for %s, %s", host, serv);

	freeaddrinfo(ressave);

	struct timeval timeout;
	timeout.tv_sec = 1200;
	timeout.tv_usec = 0;
	setsockopt(sockfd, SOL_SOCKET, SO_SNDTIMEO, &timeout, sizeof(timeout));
	setsockopt(sockfd, SOL_SOCKET, SO_RCVTIMEO, &timeout, sizeof(timeout));

	return (sockfd);
}

/* end tcp_connect */

int SendPass(int fd, char *pass)
{
	char buf[MAXLEN];
	int n;
	snprintf(buf, MAXLEN, "PASS %s\n", pass);
	if (debug)
		printf("C: %s", buf);
	Writen(fd, buf, strlen(buf));
	n = Readline(fd, buf, MAXLEN);
	buf[n] = 0;
	if (n == 0) {
		printf("read 0, exit\n");
		exit(-1);
	}
	if (debug)
		printf("S: %s", buf);
	if (memcmp(buf, "OK", 2) == 0)
		return 0;
	printf("Got %s, exit\n", buf);
	exit(-1);
}

void send_dir(int fd, char *remote_name)
{
	char buf[MAXLEN];
	int n;
	total_dirs++;
	memcpy(buf, "MKDIR ", 6);
	n = url_encode(remote_name, strlen(remote_name), buf + 6, MAXLEN - 7);
	if (n == 0) {
		printf("error when url_encode %s\n", remote_name);
		exit(-1);
	}
	buf[6 + n] = '\n';
	buf[6 + n + 1] = 0;
	if (debug)
		printf("C: %s", buf);
	Writen(fd, buf, 6 + n + 1);

	n = Readline(fd, buf, MAXLEN);
	buf[n] = 0;
	if (n == 0) {
		printf("read 0, exit\n");
		exit(-1);
	}
	if (debug)
		printf("S: %s", buf);
	if (memcmp(buf, "OK", 2) == 0)
		return;
	log_err("%s %s", remote_name, buf);
	printf("S: %s", buf);
}

void send_link(int fd, char *remote_name, char *linkto)
{
	char buf[MAXLEN];
	int n, n1, n2;
	total_links++;
	memcpy(buf, "MKLINK ", 7);
	n1 = url_encode(remote_name, strlen(remote_name), buf + 7, MAXLEN - 8);
	if (n1 == 0) {
		printf("error when url_encode %s\n", remote_name);
		exit(-1);
	}
	buf[7 + n1] = ' ';
	n1++;
	n2 = url_encode(linkto, strlen(linkto), buf + n1 + 7, MAXLEN - 8 - n1);
	if (n2 == 0) {
		printf("error when url_encode %s\n", linkto);
		exit(-1);
	}

	buf[7 + n1 + n2] = '\n';
	buf[7 + n1 + n2 + 1] = 0;
	if (debug)
		printf("C: %s", buf);
	Writen(fd, buf, 7 + n1 + n2 + 1);
	n = Readline(fd, buf, MAXLEN);
	buf[n] = 0;
	if (n == 0) {
		printf("read 0, exit\n");
		exit(-1);
	}
	if (debug)
		printf("S: %s", buf);
	if (memcmp(buf, "OK", 2) == 0)
		return;
	log_err("%s %s", remote_name, buf);
	printf("S: %s", buf);
}

void send_file(int fd, char *local_file_name, char *remote_name)
{
	char *filemd5sum;
	char buf[MAXLEN];
	int n, n1, n2;
	size_t file_size;
	struct stat buffer;
	total_files++;
	filemd5sum = file_md5sum(local_file_name);
	if (stat(local_file_name, &buffer) != 0) {
		log_err("could not stat file %s, skip\n", local_file_name);
		printf("could not stat file %s, skip\n", local_file_name);
		return;
	}
	file_size = buffer.st_size;
	total_file_len += file_size;
	n1 = snprintf(buf, MAXLEN, "FILE %s %lu ", filemd5sum, file_size);

	n2 = url_encode(remote_name, strlen(remote_name), buf + n1, MAXLEN - n1 - 1);
	if (n2 == 0) {
		printf("error when url_encode %s\n", remote_name);
		exit(-1);
	}
	buf[n1 + n2] = '\n';
	buf[n1 + n2 + 1] = 0;
	if (debug)
		printf("C: %s", buf);
	Writen(fd, buf, n1 + n2 + 1);

	n = Readline(fd, buf, MAXLEN);
	buf[n] = 0;
	if (n == 0) {
		printf("read 0, exit\n");
		exit(-1);
	}
	if (debug)
		printf("S: %s", buf);
	if (memcmp(buf, "OK", 2) == 0)
		return;
	else if (memcmp(buf, "ERROR", 5) == 0) {
		log_err("%s --> %s %s", local_file_name, remote_name, buf);
		return;
	}
	if (memcmp(buf, "DATA", 4) != 0) {
		printf("S: %s", buf);
		exit(-1);
	}

	if (debug)
		printf("I will send file\n");

	FILE *fp;
	fp = fopen(local_file_name, "r");
	if (fp == NULL) {
		printf("open file error: %s\n", local_file_name);
		exit(-1);
	}
	size_t bytes_send = 0;

	while (1) {
		size_t need_read = file_size - bytes_send;
		size_t bytes_read;
		if (need_read > 0) {
			if (need_read > MAXLEN)
				bytes_read = fread(buf, 1, MAXLEN, fp);
			else
				bytes_read = fread(buf, 1, need_read, fp);
			if (bytes_read > 0) {
				upload_file_len += bytes_read;
				Writen(fd, buf, bytes_read);
				bytes_send += bytes_read;
			} else
				break;
		} else
			break;
	}
	fclose(fp);
	n = Readline(fd, buf, MAXLEN);
	buf[n] = 0;
	if (n == 0) {
		printf("read 0, exit\n");
		exit(-1);
	}
	if (debug)
		printf("S: %s", buf);
	if (memcmp(buf, "OK", 2) == 0)
		return;
	else if (memcmp(buf, "ERROR", 5) == 0) {
		printf("S: %s", buf);
		log_err("%s --> %s %s", local_file_name, remote_name, buf);
		return;
	}
	printf("S: %s", buf);
	exit(-1);
}

void send_whole_dir(int fd, char *dir, char *remote_dir)
{
	DIR *dirp;
	struct dirent *direntp;
	dirp = opendir(dir);
	if (dirp == NULL) {
		printf("opendir error %s\n", dir);
		exit(-1);
	}
	while ((direntp = readdir(dirp)) != NULL) {
		struct stat st;
		char lfile_name[PATH_MAX];
		if (strcmp(direntp->d_name, ".") == 0)
			continue;
		if (strcmp(direntp->d_name, "..") == 0)
			continue;
		snprintf(lfile_name, PATH_MAX, "%s/%s", dir, direntp->d_name);
		if (lstat(lfile_name, &st) != 0) {
			printf("lstat error: %s\n", lfile_name);
			exit(-1);
		}
		if (S_ISDIR(st.st_mode)) {
			char buf[PATH_MAX];
			if (debug)
				printf("DIR %s\n", lfile_name);
			printf("%s\n", lfile_name);
			snprintf(buf, PATH_MAX, "%s/%s", remote_dir, direntp->d_name);
			send_dir(fd, buf);
			send_whole_dir(fd, lfile_name, buf);
		} else if (S_ISLNK(st.st_mode)) {
			char buf[PATH_MAX], lpath[PATH_MAX];
			int n;
			if (debug)
				printf("LINK %s\n", lfile_name);
			n = readlink(lfile_name, lpath, PATH_MAX - 1);
			if (n == -1) {
				printf("readlink error %s\n", lfile_name);
				exit(-1);
			}
			lpath[n] = 0;
			printf("%s LINK\n", lfile_name);
			snprintf(buf, PATH_MAX, "%s/%s", remote_dir, direntp->d_name);
			send_link(fd, buf, lpath);
		} else if (S_ISREG(st.st_mode)) {
			char buf[PATH_MAX];
			if (debug)
				printf("FILE %s\n", lfile_name);
			printf("%s\n", lfile_name);
			snprintf(buf, PATH_MAX, "%s/%s", remote_dir, direntp->d_name);
			send_file(fd, lfile_name, buf);
		} else
			printf("%s SKIP\n", lfile_name);
	}
	closedir(dirp);
}

void end_backup(int fd)
{
	char buf[MAXLEN];
	int n;
	sprintf(buf, "END\n");
	Writen(fd, buf, strlen(buf));
	n = Readline(fd, buf, MAXLEN);
	buf[n] = 0;
	if (n == 0) {
		printf("read 0, exit\n");
		exit(-1);
	}
	printf("End of backup, S: %s", buf);

	printf("Files/Dirs/Links: %zu/%zu/%zu, skipped %zu, UploadBytes/TotalBytes: %zu/%zu\n",
	       total_files, total_dirs, total_links, skipped_files,
	       upload_file_len, total_file_len);
	if (md5cache_fp)
		save_md5sum_cache();
	if (haserror) {
		printf("Encountered error when backuping file\n");
		printf("Error msg append to %s, please check it\n", error_log_file);
	}
	exit(haserror);
}

void usage(void)
{
	printf("Version: %s\n", VERSION);
	printf("Usage:\n");
	printf("./hbackup [ -d ] [ -e err_log_file ] \n"
	       "           [ -m md5cache.txt ] HostName Port Password File/DirToSend RemoteName\n");
	printf(" options:\n");
	printf("    -d              enable debug\n");
	printf
	    ("    -e err_log_file error msg will be append to err_log_file, and continue to run\n");
	printf
	    ("    -m md5cache.txt md5sum_cache will be used if the file\'s mtime does not change.\n");
	printf("                    md5sum_cache_file must be created before use\n");
	printf("\n");
	exit(-1);
}

int main(int argc, char *argv[])
{
	int c;
	int fd;

	while ((c = getopt(argc, argv, "de:m:")) != EOF)
		switch (c) {
		case 'd':
			debug = 1;
			break;
		case 'e':
			strncpy(error_log_file, optarg, PATH_MAX);
			break;
		case 'm':
			strncpy(md5cache_file, optarg, PATH_MAX);
			break;
		}

	if (argc - optind != 5)
		usage();

	strncpy(local_file_name, argv[optind + 3], PATH_MAX);
	strncpy(remote_file_name, argv[optind + 4], PATH_MAX);

	while (strlen(local_file_name) > 0 && local_file_name[strlen(local_file_name) - 1] == '/')
		local_file_name[strlen(local_file_name) - 1] = 0;
	while (strlen(remote_file_name) > 0
	       && remote_file_name[strlen(remote_file_name) - 1] == '/')
		remote_file_name[strlen(remote_file_name) - 1] = 0;
	if (debug) {
		printf("           debug = 1\n");
		printf("    exclude_regx = \n");
		printf("  error_log_file = %s\n", error_log_file);
		printf("   md5cache_file = %s\n", md5cache_file);
		printf("============================\n");
		printf("            host = %s\n", argv[optind]);
		printf("            port = %s\n", argv[optind + 1]);
		printf("            pass = %s\n", argv[optind + 2]);
		printf(" local_file_name = %s\n", local_file_name);
		printf("remote_file_name = %s\n", remote_file_name);
		printf("\n");
	}

	if (md5cache_file[0]) {
		md5cache_fp = fopen(md5cache_file, "r+");
		if (md5cache_fp == NULL) {
			printf("open file %s error, exit\n", md5cache_file);
			exit(-1);
		}
		load_md5sum_cache();
	}
	if (error_log_file[0]) {
		error_log_fp = fopen(error_log_file, "a");
		if (error_log_fp == NULL) {
			printf("open error log file %s error, exit\n", error_log_file);
			exit(-1);
		}
	}
	fd = tcp_connect(argv[optind], argv[optind + 1]);
	SendPass(fd, argv[optind + 2]);

	struct stat st;
	if (lstat(local_file_name, &st) != 0) {
		printf("lstat error: %s\n", local_file_name);
		exit(-1);
	}
	if (S_ISDIR(st.st_mode)) {
		if (debug)
			printf("DIR %s\n", local_file_name);
		printf("%s\n", local_file_name);
		send_whole_dir(fd, local_file_name, remote_file_name);
		end_backup(fd);
		exit(haserror);
	} else if (S_ISLNK(st.st_mode)) {
		char buf[PATH_MAX], lpath[PATH_MAX];
		int n;
		if (debug)
			printf("LINK %s\n", local_file_name);
		n = readlink(local_file_name, lpath, PATH_MAX - 1);
		if (n == -1) {
			printf("readlink error %s\n", local_file_name);
			exit(-1);
		}
		lpath[n] = 0;
		printf("%s LINK\n", local_file_name);
		if (remote_file_name[strlen(remote_file_name)-1] == '/')
			snprintf(buf, PATH_MAX, "%s%s", remote_file_name, basename(local_file_name));
		else
			snprintf(buf, PATH_MAX, "%s", remote_file_name);
		send_link(fd, buf, lpath);
		exit(haserror);
	} else if (S_ISREG(st.st_mode)) {
		char buf[PATH_MAX];
		if (debug)
			printf("FILE %s\n", local_file_name);
		printf("%s\n", local_file_name);
		if (remote_file_name[strlen(remote_file_name)-1] == '/')
			snprintf(buf, PATH_MAX, "%s%s", remote_file_name, basename(local_file_name));
		else
			snprintf(buf, PATH_MAX, "%s", remote_file_name);
		send_file(fd, local_file_name, buf);
		exit(haserror);
	} else {
		printf("%s SKIP\n", local_file_name);
	}
	exit(-1);
}
